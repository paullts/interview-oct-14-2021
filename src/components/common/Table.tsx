import React, { ReactNode } from 'react';
import BootStrapTable from 'react-bootstrap/Table';
import { Loading } from '../../features/loading';

interface ColumnAttributes {
  className?: string;
  colSpan?: number;
  rowSpan?: number;
}
export interface HeaderProps {
  key: string;
  headerName: string;
  attrs?: ColumnAttributes;
}
export interface RowProps {
  id: string;
  value: string | number | JSX.Element | ReactNode;
  attrs?: ColumnAttributes;
}
interface TableProps {
  header: HeaderProps[];
  rows: Array<RowProps[]> | null;
  isLoading?: boolean;
}

const NoData = ({
  colSpan,
  isLoading,
}: {
  colSpan: number;
  isLoading: boolean;
}) => {
  return (
    <tr>
      <td colSpan={colSpan} className='text-center p-4'>
        {isLoading ? <Loading /> : 'No data matching the filter'}
      </td>
    </tr>
  );
};

export default function Table(props: TableProps) {
  const { header, rows, isLoading = false } = props;
  return (
    <BootStrapTable className='table' bordered responsive size='sm'>
      <thead>
        <tr>
          {header.map((h) => (
            <th key={h.key} {...h.attrs}>
              {h.headerName}
            </th>
          ))}
        </tr>
      </thead>
      <tbody>
        {!rows ? (
          <NoData colSpan={header.length} isLoading={isLoading} />
        ) : (
          rows.map((row, idx) => (
            <tr key={idx}>
              {row.map((col) => (
                <td {...col.attrs} key={col.id}>
                  {col.value}
                </td>
              ))}
            </tr>
          ))
        )}
      </tbody>
    </BootStrapTable>
  );
}
