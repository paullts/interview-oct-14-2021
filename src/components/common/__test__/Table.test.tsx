import React from 'react';
import { cleanup, render, screen } from '@testing-library/react';
import Table, { HeaderProps, RowProps } from '../Table';

const header: HeaderProps[] = [
  {
    key: 'Confirmed',
    headerName: 'total confirmed',
    attrs: { className: 'text-center text-capitalize' },
  },
  {
    key: 'Deaths',
    headerName: 'total deaths',
    attrs: { className: 'text-center text-capitalize' },
  },
  {
    key: 'Recovered',
    headerName: 'total recovered',
    attrs: { className: 'text-center text-capitalize' },
  },
];

const rows: Array<RowProps[]> = [
  [
    { id: '123', value: '100', attrs: { className: 'text-center' } },
    {
      id: 'random-uuid',
      value: '20',
      attrs: { className: 'text-center' },
    },
    {
      id: 'random-uuid-2',
      value: '30',
      attrs: { className: 'text-center' },
    },
  ],
];

describe('renders <Table /> with header and empty rows', () => {
  let renderResult: any;
  beforeEach(() => {
    renderResult = render(<Table header={header} rows={null} />);
  });
  afterEach(cleanup);

  it('should match the snapshot', () => {
    expect(renderResult.container).toMatchSnapshot();
  });

  it('should render header cols', () => {
    expect(screen.getByText(/total confirmed/i)).toBeInTheDocument();
    expect(screen.getByText(/total deaths/i)).toBeInTheDocument();
    expect(screen.getByText(/total recovered/i)).toBeInTheDocument();
  });

  it('should render NoData', () => {
    expect(
      screen.getByText(/No data matching the filter/i)
    ).toBeInTheDocument();
  });
});

describe('renders <Table /> with specific header and rows', () => {
  beforeEach(() => {
    render(<Table header={header} rows={rows} />);
  });
  afterEach(cleanup);

  it('should render header cols', () => {
    expect(screen.queryByText('No data matching the filter')).toBeNull();
    expect(screen.getByText(/total confirmed/i)).toBeInTheDocument();
    expect(screen.getByText(/total deaths/i)).toBeInTheDocument();
    expect(screen.getByText(/total recovered/i)).toBeInTheDocument();
  });
});
