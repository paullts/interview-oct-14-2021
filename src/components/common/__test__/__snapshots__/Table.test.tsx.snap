// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <Table /> with header and empty rows should match the snapshot 1`] = `
<div>
  <div
    class="table-responsive"
  >
    <table
      class="table table table-sm table-bordered"
    >
      <thead>
        <tr>
          <th
            class="text-center text-capitalize"
          >
            total confirmed
          </th>
          <th
            class="text-center text-capitalize"
          >
            total deaths
          </th>
          <th
            class="text-center text-capitalize"
          >
            total recovered
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="text-center p-4"
            colspan="3"
          >
            No data matching the filter
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
